presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RewriteExpressionWithConstantVariable.java

```diff
@@ -0,0 +1,462 @@
+import static com.facebook.presto.common.type.BigintType.BIGINT;
+import static com.facebook.presto.common.type.DateType.DATE;
+import static com.facebook.presto.common.type.IntegerType.INTEGER;
+import static com.facebook.presto.common.type.VarcharType.VARCHAR;
+import static com.facebook.presto.expressions.LogicalRowExpressions.extractConjuncts;
+import static com.facebook.presto.spi.plan.AggregationNode.singleGroupingSet;
+import static com.facebook.presto.sql.planner.plan.ChildReplacer.replaceChildren;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static java.util.Objects.requireNonNull;
+import static java.util.function.Function.identity;
+
+/**
+ * Get constant from filter and project node. Rewrite expressions in parent nodes with the constant.
+ * For example, for query "select orderkey, orderpriority, avg(totalprice) from orders where orderpriority='3-MEDIUM' group by 1, 2", the query plan changes from
+ * <pre>
+ *     - OutputNode
+ *          orderkey, orderpriority, avg
+ *          - Aggregate
+ *              avg := avg(totalprice)
+ *              Grouping Keys := [orderkey, orderpriority]
+ *              - Filter
+ *                  orderpriority = '3-MEDIUM'
+ *                  - TableScan
+ *                      orderkey := orderkey
+ *                      orderpriority := orderpriority
+ *                      totalprice := totalprice
+ * </pre>
+ * to
+ * <pre>
+ *      - OutputNode
+ *          orderkey, expr_12, avg
+ *          - project
+ *              orderkey := orderkey
+ *              expr_12 := '3-MEDIUM'
+ *              avg := avg
+ *              - Aggregate
+ *                  avg := avg(totalprice)
+ *                  Grouping Keys := [orderkey]
+ *                  - Filter
+ *                      orderpriority = '3-MEDIUM'
+ *                      - TableScan
+ *                          orderkey := orderkey
+ *                          orderpriority := orderpriority
+ *                          totalprice := totalprice
+ * </pre>
+ */
+public class RewriteExpressionWithConstantVariable
+        implements PlanOptimizer
+{
+    private static final List<Type> SUPPORTED_JOIN_KEY_TYPE = ImmutableList.of(BIGINT, INTEGER, VARCHAR, DATE);
+
+    private final FunctionAndTypeManager functionAndTypeManager;
+    public RewriteExpressionWithConstantVariable(FunctionAndTypeManager functionAndTypeManager)
+    {
+        this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, VariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
+    {
+        if (isRewriteExpressionWithConstantEnabled(session)) {
```

Comment:
can you log when this optimizer is invoked?

Reply:
I think it should already be covered by Lyublena's PR https://github.com/prestodb/presto/pull/19772